File: leetcode/dp/l55.ts
function canJump(nums: number[]): boolean {
    let step = 0
    for (let i = 0; i < nums.length; i++) {
        if (i > step) return false;
        step = Math.max(step, i + nums[i])
    }
    return true;
};

canJump([3, 2, 1, 0, 4])

File: src/l55.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
    bool canJump_tanxin(std::vector<int>& nums) {
        int step = 0;
        for (int i = 0; i < nums.size();i ++) {
            if (i > step) return false;
            step = std::max(step, i + nums[i]);
        }
        return true;
    }
    // reverse
    // 只要记录最后有地方能通过就行
    bool canJump(std::vector<int>& nums) {
        int pos = nums.size() - 1;
        for (int i = nums.size() - 2; i >= 0;i--) {
            if (i + nums[i] >= pos) {
                pos = i;
            }
        }
        return pos == 0;
    }
};